app/com/lucidchart/open/cashy/controllers/UploadController.scala

```diff
@@ -0,0 +1,142 @@
+    })
+  )
+
+  /**
+   * Upload form, authentication required
+   */
+  def index = AuthAction.authenticatedUser { implicit user =>
+    Action {  implicit request =>
+      Ok(views.html.upload.index(uploadForm, buckets)(request,userOption=Some(user)))
+    }
+  }
+
+  def uploadToS3 = AuthAction.authenticatedUser { implicit user =>
+    Action(parse.multipartFormData(FileHandler.handleFilePartAsByteArray)) { implicit request =>
+
+      val fileOption = request.body.file("assetFile")
+      val formWithData = uploadForm.bindFromRequest()
+      val formWithFileValidation = fileOption match {
+        case None => {
+          formWithData.copy(errors = formWithData.errors :+ new FormError("assetFile", "Must select a file"))
+        }
+        case Some(file) => formWithData
+      }
+
+
+      formWithFileValidation.fold(
+        formWithErrors => {
+          Ok(views.html.upload.index(formWithErrors, buckets)(request,userOption=Some(user)))
+        },
+        data => {
+
+          try {
+
+            val bucket = data.bucket
+            val assetName = data.assetName
+
+            // Get the file bytes and content type
+            val (bytes, contentType) = fileOption.get match {
+              case FilePart(key, filename, contentType, bytes) => {
+                (bytes, contentType)
+              }
+            }
+
+            // Check if gzip compression is appropriate
+            val compressedBytes = GzipHelper.compress(bytes)
+            val gzipSavings = (bytes.length - compressedBytes.length) / bytes.length.toDouble
+            val gzipUploaded = if (gzipSavings >= minGzipSavings) {
+              if(!S3Client.uploadToS3(bucket, assetName, compressedBytes, contentType, gzipped=true)) {
+                throw new UploadFailedException("Uploading gzip version for file failed: " + (bucket, assetName))
+              }
+              true
+            } else {
+              false
+            }
+
+            if(S3Client.uploadToS3(bucket, assetName, bytes, contentType)) {
+
+              AssetModel.createAsset(bucket, assetName, user.id)
+
+              val assetLink = bucketCloudfrontMap.get(bucket).get + assetName
+
+              AuditModel.createUploadAudit(user.id, bucket, assetName, assetLink, gzipUploaded)
+
+              Ok(views.html.upload.complete(assetLink)(request,userOption=Some(user)))
+            } else {
+              // If gzip upload happened but this one failed, have to delete the gzip one
+              if (gzipUploaded) {
+                try {
+                  S3Client.removeFromS3(bucket, assetName, gzipped=true)
+                }
+                catch {
+                  case e: Exception => {
+                    throw new UploadFailedException("Uploading non-gzip version for file failed and could not delete the gzip version.  Please contact your S3 administrator to correct the issue: " + (bucket, assetName))
```

Comment:
This was the solution for problems deleting from S3.  Figure it's best if they contact someone with knowledge of S3 and real access rather than try a few times without knowing exactly why it failed.